pkg/kallsyms/kallsyms.go

```diff
@@ -149,20 +149,31 @@ var (
 //
 //	kallsyms.SpecUpdateAddresses(spec, []string{"socket_file_ops"})
 func SpecUpdateAddresses(spec *ebpf.CollectionSpec, symbols []string) error {
-	kAllSymsFactory := NewKAllSyms
-	return specUpdateAddresses(kAllSymsFactory, spec, symbols)
+	if len(symbols) == 0 {
+		// Nothing to do
+		return nil
+	}
+
+	kAllSyms, err := NewKAllSyms()
+	if err != nil {
+		return err
+	}
```

Comment:
Before this change, `NewKAllSyms()` was only called if we have missing entries in the cache. With this change, we lose performance by reading and parsing /proc/kallsyms every time `SpecUpdateAddresses()` is called.

I suspect this causes the following performance issue in trace-dns and top-ebpf:
https://github.com/inspektor-gadget/inspektor-gadget/commit/661af6b62c392162aa18f5740645b784e8baa206#commitcomment-139171154

Reply:
Great catch, you are right it seem to be related:

```
→ go test -exec sudo  -bench='BenchmarkAllGadgetsWithContainers/container0/top-ebpf'  -run=Benchmark ./internal/benchmarks/...
goos: linux
goarch: amd64
pkg: github.com/inspektor-gadget/inspektor-gadget/internal/benchmarks
cpu: Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
BenchmarkAllGadgetsWithContainers/container0/top-ebpf-8                        2         515444363 ns/op
PASS
ok      github.com/inspektor-gadget/inspektor-gadget/internal/benchmarks        1.716s

```

compared to using `sync.Once` as recommend [here](https://github.com/inspektor-gadget/inspektor-gadget/pull/2280#discussion_r1502323160):

```
→ go test -exec sudo  -bench='BenchmarkAllGadgetsWithContainers/container0/top-ebpf'  -run=Benchmark ./internal/benchmarks/...
goos: linux
goarch: amd64
pkg: github.com/inspektor-gadget/inspektor-gadget/internal/benchmarks
cpu: Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
BenchmarkAllGadgetsWithContainers/container0/top-ebpf-8                      100          11858637 ns/op
PASS
ok      github.com/inspektor-gadget/inspektor-gadget/internal/benchmarks        1.905s

```